File: compound.cpp
#include <iostream>
#include <H5Cpp.h>
#include <vector>
#include <cstring>  // For memcpy

using namespace H5;

const H5std_string FILE_NAME("test.h5");
const H5std_string DATASET_NAME("Demand");
const H5std_string ATTRIBUTE_NAME("GIT root revision");

// ✅ Define the packed structure
struct __attribute__((packed)) Shipment {
    uint64_t shipmentId;
    char origCountry[2];
    char origSlic[5];
    uint8_t origSort;
    char destCountry[2];
    char destSlic[5];
    uint8_t destIbi;
    char destPostalCode[9];
    char shipper[10];
    uint8_t service;
    uint8_t packageType;
    uint8_t accessorials;
    uint16_t pieces;
    uint16_t weight;
    uint32_t cube;
    uint8_t committedTnt;
    uint8_t committedDate;
};

// Number of records
const int NUM_RECORDS = 1750;

int main() {
    try {
        // Create a new HDF5 file
        H5File file(FILE_NAME, H5F_ACC_TRUNC);

        // ✅ Define the compound data type directly from `Shipment`
        CompType compoundType(sizeof(Shipment));
        compoundType.insertMember("shipmentId", HOFFSET(Shipment, shipmentId), PredType::NATIVE_UINT64);
        compoundType.insertMember("origCountry", HOFFSET(Shipment, origCountry), StrType(PredType::C_S1, 2));
        compoundType.insertMember("origSlic", HOFFSET(Shipment, origSlic), StrType(PredType::C_S1, 5));
        compoundType.insertMember("origSort", HOFFSET(Shipment, origSort), PredType::NATIVE_UINT8);
        compoundType.insertMember("destCountry", HOFFSET(Shipment, destCountry), StrType(PredType::C_S1, 2));
        compoundType.insertMember("destSlic", HOFFSET(Shipment, destSlic), StrType(PredType::C_S1, 5));
        compoundType.insertMember("destIbi", HOFFSET(Shipment, destIbi), PredType::NATIVE_UINT8);
        compoundType.insertMember("destPostalCode", HOFFSET(Shipment, destPostalCode), StrType(PredType::C_S1, 9));
        compoundType.insertMember("shipper", HOFFSET(Shipment, shipper), StrType(PredType::C_S1, 10));
        compoundType.insertMember("service", HOFFSET(Shipment, service), PredType::NATIVE_UINT8);
        compoundType.insertMember("packageType", HOFFSET(Shipment, packageType), PredType::NATIVE_UINT8);
        compoundType.insertMember("accessorials", HOFFSET(Shipment, accessorials), PredType::NATIVE_UINT8);
        compoundType.insertMember("pieces", HOFFSET(Shipment, pieces), PredType::NATIVE_UINT16);
        compoundType.insertMember("weight", HOFFSET(Shipment, weight), PredType::NATIVE_UINT16);
        compoundType.insertMember("cube", HOFFSET(Shipment, cube), PredType::NATIVE_UINT32);
        compoundType.insertMember("committedTnt", HOFFSET(Shipment, committedTnt), PredType::NATIVE_UINT8);
        compoundType.insertMember("committedDate", HOFFSET(Shipment, committedDate), PredType::NATIVE_UINT8);

        // ✅ Create data space
        hsize_t dim[1] = { NUM_RECORDS };
        hsize_t maxdim[1] = { NUM_RECORDS }; // Explicitly fix maxdims
        DataSpace dataspace(1, dim, maxdim);

        // ✅ Create dataset
        DataSet dataset = file.createDataSet(DATASET_NAME, compoundType, dataspace);

        // ✅ ADD ATTRIBUTE: "GIT root revision"
        H5std_string attribute_value = "Revision: , URL: ";
        StrType attr_type(PredType::C_S1, attribute_value.size());
        DataSpace attr_space(H5S_SCALAR);
        Attribute attribute = dataset.createAttribute(ATTRIBUTE_NAME, attr_type, attr_space);
        attribute.write(attr_type, attribute_value);
        attribute.close();

        // ✅ Allocate a vector of `Shipment` structs
        std::vector<Shipment> shipments(NUM_RECORDS);

        // ✅ Fill the struct array
        for (int i = 0; i < NUM_RECORDS; i++) {
            shipments[i].shipmentId = i + 1000;
            memcpy(shipments[i].origCountry, "US", 2);
            memcpy(shipments[i].origSlic, "12345", 5);
            shipments[i].origSort = 4;
            memcpy(shipments[i].destCountry, "CA", 2);
            memcpy(shipments[i].destSlic, "67890", 5);
            shipments[i].destIbi = 0;
            memcpy(shipments[i].destPostalCode, "A1B2C3", std::min(sizeof("A1B2C3") - 1, sizeof(shipments[i].destPostalCode)));
            memcpy(shipments[i].shipper, "FedEx", std::min(sizeof("FedEx") - 1, sizeof(shipments[i].shipper)));
            shipments[i].service = 0;
            shipments[i].packageType = 3;
            shipments[i].accessorials = 0;
            shipments[i].pieces = 2;
            shipments[i].weight = 50;
            shipments[i].cube = 1200;
            shipments[i].committedTnt = 255;
            shipments[i].committedDate = 3;
        }

        // ✅ Write to dataset
        dataset.write(shipments.data(), compoundType);

        // ✅ Close resources
        dataset.close();
        file.close();

        std::cout << "HDF5 file created and written successfully!\n";
    } catch (Exception &e) {
        std::cerr << "HDF5 Exception: " << e.getCDetailMsg() << std::endl;
        return 1;
    }

    return 0;
}

File: compoundexamples/compound.py
import h5py
import numpy as np

FILE_NAME = "testpy.h5"
DATASET_NAME = "Demand"
NUM_RECORDS = 1750
STRUCT_SIZE = 56
buffer_size = NUM_RECORDS * STRUCT_SIZE
buffer = bytearray(buffer_size)

# ✅ Define structured NumPy dtype (Mimics C++ Packed Struct)
structured_dtype = np.dtype([
    ("shipmentId", "<u8"),
    ("origCountry", "S2"),
    ("origSlic", "S5"),
    ("origSort", "S1"),
    ("destCountry", "S2"),
    ("destSlic", "S5"),
    ("destIbi", "S1"),
    ("destPostalCode", "S9"),
    ("shipper", "S10"),
    ("service", "S1"),
    ("packageType", "S1"),
    ("accessorials", "S1"),
    ("pieces", "<u2"),
    ("weight", "<u2"),
    ("cube", "<u4"),
    ("committedTnt", "S1"),
    ("committedDate", "S1"),
])

with h5py.File(FILE_NAME, "w") as hdf_file:
    # ✅ Create dataset with compound type using high-level interface
    dset = hdf_file.create_dataset(
        DATASET_NAME,
        shape=(NUM_RECORDS,),
        dtype=structured_dtype  # Use the NumPy dtype directly
    )
    
    # ✅ ADD ATTRIBUTE: "GIT root revision" with fixed-length string
    attribute_name = "GIT root revision"
    attribute_value = "Revision: , URL: "
    # Create a fixed-length string datatype matching the value length
    fixed_str_dtype = h5py.string_dtype(encoding='ascii', length=len(attribute_value))
    # Create the attribute with the fixed-length type
    dset.attrs.create(attribute_name, attribute_value, dtype=fixed_str_dtype)

    # ✅ Populate buffer
    for i in range(NUM_RECORDS):
        offset = i * STRUCT_SIZE

        buffer[offset : offset + 8] = (1000 + i).to_bytes(8, "little")
        buffer[offset + 8 : offset + 10] = b"US"
        buffer[offset + 10 : offset + 15] = b"12345"
        buffer[offset + 15 : offset + 16] = b"L"
        buffer[offset + 16 : offset + 18] = b"CA"
        buffer[offset + 18 : offset + 23] = b"67890"
        buffer[offset + 23 : offset + 24] = b"N"
        buffer[offset + 24 : offset + 33] = b"A1B2C3".ljust(9, b"\x00")  # Ensure exact 9 bytes
        buffer[offset + 33 : offset + 43] = b"FedEx".ljust(10, b"\x00")  # Ensure exact 10 bytes
        buffer[offset + 43 : offset + 44] = b"E"
        buffer[offset + 44 : offset + 45] = b"B"
        buffer[offset + 45 : offset + 46] = b"N"
        buffer[offset + 46 : offset + 48] = (2).to_bytes(2, "little")
        buffer[offset + 48 : offset + 50] = (50).to_bytes(2, "little")
        buffer[offset + 50 : offset + 54] = (1200).to_bytes(4, "little")
        buffer[offset + 54 : offset + 55] = b"Y"
        buffer[offset + 55 : offset + 56] = b"D"

    # ✅ Convert buffer to structured NumPy array
    print(f"First Record Raw Bytes (i=0): {buffer[:STRUCT_SIZE].hex()}")  # 🚀 Debug Dump
    numpy_data = np.frombuffer(buffer, dtype=structured_dtype)

    # ✅ Write data to dataset
    dset[()] = numpy_data

print(f"HDF5 file '{FILE_NAME}' created successfully with dynamically built compound type.")

# Verify attribute
with h5py.File(FILE_NAME, "r") as hdf_file:
    dataset = hdf_file[DATASET_NAME]
    print(f"Attribute '{attribute_name}': {dataset.attrs[attribute_name]}")